include/boost/url/detail/impl/params_encoded_iterator_impl.ipp

```diff
@@ -0,0 +1,205 @@
+#include <boost/assert.hpp>
+
+namespace boost {
+namespace urls {
+namespace detail {
+
+void
+params_encoded_iterator_impl::
+scan() noexcept
+{
+    string_view s(pos_, end_ - pos_);
+    std::size_t i;
+    if(pos_ != begin_ || i_ != 0)
+    {
+        BOOST_ASSERT(
+            s.starts_with('&'));
+        i = s.find_first_of('&', 1);
+    }
+    else
+    {
+        i = s.find_first_of('&');
+    }
+    if( i == string_view::npos)
+        i = s.size();
+    nk_ = string_view(
+        pos_, i).find_first_of('=');
+    if(nk_ != string_view::npos)
+    {
+        nv_ = i - nk_;
+    }
+    else
+    {
+        // has_value==false
+        nk_ = i;
+        nv_ = 0;
+    }
+}
+
+params_encoded_iterator_impl::
+params_encoded_iterator_impl(
+    string_view s) noexcept
+    : begin_(s.data())
+    , pos_(s.data())
+    , end_(s.data() + s.size())
+{
+    scan();
+}
+
+params_encoded_iterator_impl::
+params_encoded_iterator_impl(
+    string_view s,
+    std::size_t nparam,
+    int) noexcept
+    : i_(nparam)
+    , begin_(s.data())
+    , pos_(s.data() + s.size())
+    , end_(s.data() + s.size())
+{
+}
+
+void
+params_encoded_iterator_impl::
+increment() noexcept
+{
+    // valid domain
+    BOOST_ASSERT(begin_ != nullptr);
+    BOOST_ASSERT(end_ != nullptr);
+
+    // cannot increment one past end
+    BOOST_ASSERT(pos_ != end_);
+
+    ++i_;
+    if (pos_ != nullptr)
+    {
+        pos_ += nk_ + nv_;
+        if(pos_ == end_)
+        {
+            nk_ = 0;
+            nv_ = 0;
+            return;
+        }
+        scan();
+        return;
+    }
+    else
+    {
+        pos_ = begin_;
+    }
+    scan();
+}
+
+void
+params_encoded_iterator_impl::
+decrement() noexcept
+{
+    // valid domain
+    BOOST_ASSERT(begin_ != nullptr);
+    BOOST_ASSERT(end_ != nullptr);
+
+    // cannot decrement one before begin
+    BOOST_ASSERT(pos_ != nullptr);
+
+    --i_;
+    if (begin_ != pos_)
+    {
+        string_view s(begin_, pos_ - begin_);
+        std::size_t i = s.find_last_of('&');
+        pos_ = begin_;
+        if (i != string_view::npos)
+            pos_ = begin_ + i;
+        scan();
+    }
+    else
+    {
+        // one before begin
+        pos_ = nullptr;
+    }
+}
+
+string_view
+params_encoded_iterator_impl::
+encoded_key() const noexcept
+{
+    // valid domain
+    BOOST_ASSERT(begin_ != nullptr);
+    BOOST_ASSERT(end_ != nullptr);
+
+    // cannot dereference one past end
+    BOOST_ASSERT(pos_ != end_);
+
+    // cannot dereference one before begin
+    BOOST_ASSERT(pos_ != nullptr);
+
+    if(pos_ != begin_)
+        return string_view(
+            pos_ + 1, nk_ - 1);
+    return string_view{ pos_, nk_ };
+}
+
+auto
+params_encoded_iterator_impl::
+dereference() const ->
+    query_param_view
+{
+    // valid domain
+    BOOST_ASSERT(begin_ != nullptr);
+    BOOST_ASSERT(end_ != nullptr);
+
+    // cannot dereference one past end
+    BOOST_ASSERT(pos_ != end_);
+
+    // cannot dereference one before begin
```

Comment:
one before begin? what is that?

Reply:
std::prev(p.begin())

Reply:
Wait, no,... the iterator should never be in the one-before-begin state